source/date_time.c

```diff
@@ -0,0 +1,826 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <aws/common/date_time.h>
+
+#include <aws/common/byte_buf.h>
+#include <aws/common/clock.h>
+#include <aws/common/time.h>
+
+#include <ctype.h>
+#include <math.h>
+
+static const char *RFC822_DATE_FORMAT_STR_MINUS_Z = "%a, %d %b %Y %H:%M:%S GMT";
+static const char *RFC822_DATE_FORMAT_STR_WITH_Z = "%a, %d %b %Y %H:%M:%S %Z";
+static const char *ISO_8601_LONG_DATE_FORMAT_STR = "%Y-%m-%dT%H:%M:%SZ";
+
+static int get_week_day_number_from_str(const char *time_string, size_t startIndex, size_t stopIndex) {
+    if (stopIndex - startIndex < 3) {
+        return -1;
+    }
+
+    size_t index = startIndex;
+
+    char c = time_string[index];
+    char next = 0;
+
+    /* it's ugly but this should compile down to EXACTLY 3 comparisons and no memory allocations */
+    switch (c) {
+        case 'S':
+        case 's':
+            next = time_string[++index];
+            switch (next) {
+                case 'A':
+                case 'a':
+                    next = time_string[++index];
+                    switch (next) {
+                        case 'T':
+                        case 't':
+                            return 6;
+                        default:
+                            return -1;
+                    }
+                case 'U':
+                case 'u':
+                    next = time_string[++index];
+                    switch (next) {
+                        case 'N':
+                        case 'n':
+                            return 0;
+                        default:
+                            return -1;
+                    }
+                default:
+                    return -1;
+            }
+        case 'T':
+        case 't':
+            next = time_string[++index];
+            switch (next) {
+                case 'H':
```

Comment:
Despite the comment above, this nested switch may require up to nine comparisons:

* 4 comparisons for the outer switch (as a binary search of 10 elements)
* 3 comparisons for the middle switch (as a binary search of 4 elements)
* Two comparisons for the innermost switch

It could also be implemented as three switch tables, or some combination of switchtables and binary searches, at the expense of L1, but exactly three comparisons is an unlikely compilation result.

Anyway - writing this kind of code by hand is a bit of a waste of effort, and quite hard to debug. I would recommend one of:

* using a simple strcmp sequence
* doing a hash table lookup
* using a lexer generator like re2c
* using a perfect hash generator like gperf, if you really want the maximum possible performance

Reply:
For matching a fixed set of strings, strcmp is way simpler:

```
char day_buf[4] = {0};
strncpy(day_buf, time_string, sizeof(day_buf) - 1);
for (size_t i = 0; i < 3; i++) day_buf[i] = tolower(day_buf[i]);

if (!strcmp(time_string, "mon")) return 1;
if (!strcmp(time_string, "tue")) return 2;
// ...
```

Reply:
Yeah, I think it's often simpler inside a single branch of a state machine, just not for the whole thing. I'll see what I can do on it.